STRATCONN-5938 - [RoadwayAI] - New Destination

mordente opened the pull request:
## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [If destination is already live] Tested for backward compatibility of destination. **Note:** New required fields are a breaking change.
- [ ] [Segmenters] Tested in the staging environment
- [ ] [Segmenters] [If applicable for this change] Tested for regression with Hadron.

mordente commented:
> hi @mordente thanks for the PR for a new Integration! I'll review it in the next 24h. Kind regards, Joe

Hey Joe. Hope you're doing well.
How we looking on this PR?

joe-ayoub-segment reviewed (changes requested):
Hi @mordente, 
Thanks again for the PR and apologies for the slow response time. 

I reviewed the code, and while I didn't see any major code issues, I do think that the fields could be better designed. 

Would it be possible for you to create Typescript Types for each of the 4 types of requests you want to send to your platform?

This will clarify what the request JSON needs to look like. 

After that, I think we should merge some of the individual fields into object fields. An object field is just a way to collate individual fields together in the UI. You can separate them back into individual key:value pairs in the perform() and performBatch() functions afterwards if needed. 

You are already using an object field for the utm_properties. 

Perhaps you could group your fields into separate object fields for: 
 - page properties
 - user traits / properties
 - mobile device properties
 - mobile app properties
 - etc...?

Lastly, I think it would be useful to define a single file to hold all of your fields. 
Then pull in the fields as needed into each of your Actions. 

Let me know if you have any questions about this. 
Best regards,
Joe

joe-ayoub-segment commented:
Hi again @mordente,
The unit tests are also failing. 

![image](https://github.com/user-attachments/assets/372a8c60-2bcd-461d-b730-75080bce273e)

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/unified-fields.ts:
```diff
@@ -0,0 +1,451 @@
+  description: 'UTM tracking and campaign attribution properties',
+  properties: {
+    utm_source: {
+      label: 'UTM Source',
```
please add a description

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/unified-fields.ts:
```diff
@@ -0,0 +1,451 @@
+      type: 'string'
+    },
+    utm_medium: {
+      label: 'UTM Medium',
```
please add a description

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/unified-fields.ts:
```diff
@@ -0,0 +1,451 @@
+      type: 'string'
+    },
+    utm_campaign: {
+      label: 'UTM Campaign',
```
please add a description

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/unified-fields.ts:
```diff
@@ -0,0 +1,451 @@
+      type: 'string'
+    },
+    utm_term: {
+      label: 'UTM Term',
```
please add a description

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/unified-fields.ts:
```diff
@@ -0,0 +1,451 @@
+      type: 'string'
+    },
+    utm_content: {
+      label: 'UTM Content',
```
please add a description

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/payload-transformer.ts:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Utility functions to transform object fields back to individual key:value pairs
+ * for RoadwayAI API compatibility
+ */
+
+/**
+ * Transforms a payload with object fields into a flattened payload with individual key:value pairs
+ * @param payload The payload containing object fields
+ * @returns Flattened payload with individual properties
+ */
+export function flattenPayload(payload: any): any {
+  const { app_properties, location_properties, page_properties, utm_properties, ...otherFields } = payload
+
+  // Flatten the nested objects into individual properties
+  return {
+    ...otherFields,
+    // Spread app properties as individual fields (app_name, app_version)
+    ...(app_properties || {}),
+    // Spread location properties as individual fields (country, region, language)
+    ...(location_properties || {}),
+    // Spread page properties as individual fields (url, referrer)
+    ...(page_properties || {}),
+    // Spread UTM properties as individual fields (utm_source, utm_medium, etc.)
+    ...(utm_properties || {})
+  }
+}
+
+/**
+ * Transforms an array of payloads for batch operations
+ * @param payloads Array of payloads containing object fields
+ * @returns Array of flattened payloads with individual properties
+ */
+export function flattenPayloadBatch(payloads: any[]): any[] {
+  return payloads.map(flattenPayload)
+}
```
```suggestion
import { Payload as TrackPayload } from './trackEvent/generated-types'
import { Payload as IdentifyUserPayload } from './identifyUser/generated-types'
import { Payload as GroupUserPayload } from './groupUser/generated-types'
import { Payload as TrackPageViewPayload } from './trackPageView/generated-types'

/**
 * Utility functions to transform object fields back to individual key:value pairs
 * for RoadwayAI API compatibility
 */

type AllPayloads = TrackPayload | IdentifyUserPayload | GroupUserPayload | TrackPageViewPayload

/**
 * Transforms a payload with object fields into a flattened payload with individual key:value pairs
 * @param payload The payload containing object fields
 * @returns Flattened payload with individual properties
 */
export function flattenPayload<T>(payload: AllPayloads ): T {
    const {
    app_properties,
    location_properties,
    page_properties,
    utm_properties,
    ...otherFields
  } = payload as Partial<AllPayloads> & Record<string, unknown>

  // Flatten the nested objects into individual properties
  return {
    ...otherFields,
    // Spread app properties as individual fields (app_name, app_version)
    ...(app_properties || {}),
    // Spread location properties as individual fields (country, region, language)
    ...(location_properties || {}),
    // Spread page properties as individual fields (url, referrer)
    ...(page_properties || {}),
    // Spread UTM properties as individual fields (utm_source, utm_medium, etc.)
    ...(utm_properties || {})
  } as T
}

/**
 * Transforms an array of payloads for batch operations
 * @param payloads Array of payloads containing object fields
 * @returns Array of flattened payloads with individual properties
 */
export function flattenPayloadBatch<T>(payloads: AllPayloads[]): T[] {
  return payloads.map(flattenPayload) as T[]
}
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/groupUser/index.ts:
```diff
@@ -0,0 +1,37 @@
+  },
+
+  performBatch: async (request, { settings, payload }) => {
+    const transformedPayloads = flattenPayloadBatch(payload) as GroupUserRequest[]
```
```suggestion
    const transformedPayloads = flattenPayloadBatch<GroupUserRequest[]>(payload)
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/identifyUser/index.ts:
```diff
@@ -0,0 +1,37 @@
+  },
+
+  performBatch: async (request, { settings, payload }) => {
+    const transformedPayloads = flattenPayloadBatch(payload) as IdentifyUserRequest[]
```
```suggestion
    const transformedPayloads = flattenPayloadBatch<IdentifyUserRequest[]>(payload)
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/trackEvent/index.ts:
```diff
@@ -0,0 +1,38 @@
+  defaultSubscription: 'type = "track"',
+  fields: getTrackEventFields(),
+  perform: async (request, { settings, payload }) => {
+    const flattenedPayload = flattenPayload(payload) as TrackEventRequest
```
```suggestion
    const flattenedPayload = flattenPayload<TrackEventRequest>(payload)
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/trackPageView/index.ts:
```diff
@@ -0,0 +1,37 @@
+  defaultSubscription: 'type = "page"',
+  fields: getTrackPageViewFields(),
+  perform: async (request, { settings, payload }) => {
+    const flattenedPayload = flattenPayload(payload) as TrackPageViewRequest
```
```suggestion
    const flattenedPayload = flattenPayload<TrackPageViewRequest>(payload)
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/roadwayai/unified-fields.ts:
```diff
@@ -0,0 +1,451 @@
+    default: true
+  },
+
+  distinct_id: {
+    label: 'Distinct ID',
+    type: 'string',
+    description: 'A distinct ID specified by you.',
+    default: {
+      '@if': {
+        exists: { '@path': '$.userId' },
+        then: { '@path': '$.userId' },
+        else: { '@path': '$.anonymousId' }
+      }
+    }
+  },
+
+  anonymous_id: {
+    label: 'Anonymous ID',
+    type: 'string',
+    description: 'A distinct ID randomly generated prior to calling identify.',
+    default: {
+      '@path': '$.anonymousId'
+    }
+  },
+
+  user_id: {
+    label: 'User ID',
+    type: 'string',
+    description: 'The distinct ID after calling identify.',
+    default: {
+      '@path': '$.userId'
+    }
+  },
```
Hi @mordente quick question about the identifiers. Apologies if you already told me. 
Why are there separate fields for distinct id, userId and anonymousId?

joe-ayoub-segment reviewed (changes requested):
Hi @mordente thanks for the updates to the PR. 
It looks much better. 
I updated a few types and added some other minor suggestions but other than that and the tests it looks good to me. 

The tests are still failing. 
To run them use this command from the command line at the root of the project.  

`yarn cloud test packages/destination-actions/src/destinations/roadwayai --updateSnapshot`

Best regards,
Joe

mordente commented:
> `yarn cloud test packages/destination-actions/src/destinations/roadwayai --updateSnapshot`

Hey @joe-ayoub-segment, thanks for the review again. Just addressed those changes and was able to successfully run tests now 😄.
Also removed the distinct_id field, wasn't necessary since we have both user_id and anonymous_id.

joe-ayoub-segment commented:
Raised internal PR [here](https://github.com/segmentio/action-destinations/pull/3003) so I could update types file. 
Closing this PR.
